staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -316,6 +337,17 @@ func (s preparedGenericAPIServer) NonBlockingRun(stopCh <-chan struct{}) error {
 	return nil
 }
 
+// WaitHandler waits for all requests to finish, which is bounded by the RequestTimeout variable.
+func (s preparedGenericAPIServer) WaitHandler() {
+	for {
+		if atomic.LoadInt64(s.HandlerCounter) == 0 {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
```

Comment:
can we hide all the synchronization logic behind an abstract waiting group interface?

Reply:
You mean we implement our waitGroup `Add` and `Wait`?

Reply:
yes

Reply:
What about we call wg.Add(1) right after wg is declared.
And an extra wg.Done() before wg.Wait() ?

@sttts

Reply:
wg.Add() and wg.Done are go-routine safe. 
We only need to make sure wg.Add() is called before wg.Wait().

Reply:
Sounds Good.  It's viable.

Reply:
Only a little strange to readers.

Reply:
@CaoShuFeng I thought again, it is not viable. Because even we call wg.Done before wg.Wait, we still can not guarantee when we call wg.Wait no new wg.Add will be called.

Reply:
> Only a little strange to readers.

I think so. :)